src/optimization/steepest_descent_method.cpp

```diff
@@ -0,0 +1,292 @@
+#include "k52/optimization/steepest_descent_method.h"
+
+#ifdef BUILD_WITH_MPI
+
+#include <boost/mpi.hpp>
+#include <k52/parallel/mpi/constants.h>
+
+#endif
+
+#include <iostream>
+#include <cmath>
+
+#include <k52/optimization/params/i_continuous_parameters.h>
+
+namespace
+{
+/// @brief Golden ratio value for GoldenSectionSearch method
+/// @link https://en.wikipedia.org/wiki/Golden_ratio
+	const double kAlpha = (1.0 + sqrt(5.0)) / 2.0;
+}
+
+using ::std::vector;
+
+namespace k52
+{
+namespace optimization
+{
+
+SteepestDescentMethod::SteepestDescentMethod(
+    double increment_of_the_argument,
+    size_t max_iteration_number,
+    double precision)
+    : increment_of_the_argument_(increment_of_the_argument)
+    , max_iteration_number_(max_iteration_number)
+    , precision_(precision)
+{
+}
+
+SteepestDescentMethod* SteepestDescentMethod::Clone() const
+{
+    return new SteepestDescentMethod(
+        increment_of_the_argument_,
+        max_iteration_number_,
+        precision_);
+}
+
+std::string SteepestDescentMethod::get_name() const
+{
+    return "Steepest Descent Method";
+}
+
+
+#ifdef BUILD_WITH_MPI
+void SteepestDescentMethod::Send(boost::mpi::communicator* communicator, int target) const
+ {
+    communicator->send(target, k52::parallel::mpi::constants::kCommonTag, increment_of_the_argument_);
+    communicator->send(target, k52::parallel::mpi::constants::kCommonTag, max_iteration_number_);
+    communicator->send(target, k52::parallel::mpi::constants::kCommonTag, precision_);
+}
+
+void SteepestDescentMethod::Receive(boost::mpi::communicator* communicator, int source)
+{
+    communicator->recv(source, k52::parallel::mpi::constants::kCommonTag, increment_of_the_argument_);
+    communicator->recv(source, k52::parallel::mpi::constants::kCommonTag, max_iteration_number_);
+    communicator->recv(source, k52::parallel::mpi::constants::kCommonTag, precision_);
+}
+#endif
+
+
+vector<double> SteepestDescentMethod::FindOptimalParameters(const vector<double>& initial_parameters)
+{
+    vector<double> parameters(initial_parameters);
+    size_t counter = 0;
+    double lambda = 1; 
+
+    vector<double> next_step_parameters(parameters);
+    do
+    {
+        parameters = next_step_parameters;
+        counter++;
+        vector<double> gradient = CalculateGradient(parameters);
+        vector<double> invert_gradiend = InvertVector(gradient);
+
+        lambda = GoldenSectionSearch(parameters, lambda, invert_gradiend);
```

Comment:
For me it looks very strange that the lambda value that is found as an optimal step for exact gradient direction on previous iteration is passed as an initial step for a one dimensional search method. Why this is not a separate parameter? Is there any proof link that with an argumentation to do so?